engine/common/distributor/distributor.go

```diff
@@ -0,0 +1,175 @@
+// The consumer will be notified whenever a new event is distributed by the AsyncEventDistributor.
+// Registering a consumer can be done at any time, even after the AsyncEventDistributor has been started.
+// However, the consumer will not be notified of events that were submitted before the consumer was registered.
+func (a *AsyncEventDistributor[T]) RegisterConsumer(consumer EventConsumer[T]) {
+	a.consumerLock.Lock()
+	defer a.consumerLock.Unlock()
+
+	a.consumers = append(a.consumers, consumer)
+}
+
+// distributorWorker is a worker that is spawned by the AsyncEventDistributor to pick up events from the message store and
+// distribute them to all the consumers.
+// The worker is blocked on the distributor's notifier channel, and wakes up whenever a new event is received.
+// On waking up, the worker keeps distributing events till the message store is empty.
+func (a *AsyncEventDistributor[T]) distributorWorker(ctx irrecoverable.SignalerContext) {
+	for {
+		select {
+		case <-ctx.Done():
+			a.log.Debug().Msg("distributing event worker terminated")
+			return
+		case <-a.notifier.Channel():
+			a.distributeEvents(ctx)
+		}
+	}
+}
+
+// distributeEvents is part of the worker's logic. It keeps distributing events till the message store is empty.
+func (a *AsyncEventDistributor[T]) distributeEvents(ctx irrecoverable.SignalerContext) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			msg, ok := a.store.Get()
+			if !ok {
+				a.log.Trace().Msg("no more events to distribute, returning")
+				return
+			}
+			event, ok := msg.Payload.(T)
+			if !ok {
+				a.log.Fatal().Msgf("invalid event type expected: %T", msg.Payload)
+			}
+			lg := a.log.With().
+				Str("payload", fmt.Sprintf("%v", msg.Payload)).Logger()
+			lg.Trace().Msg("distributing event to consumers started")
+			a.distributeEvent(event)
+			lg.Trace().Msg("event distributed to all registered consumers")
+		}
+	}
+}
+
+// distributeEvent sends the event to all registered consumers. It is concurrency safe and blocks until all consumers have processed the event.
+func (a *AsyncEventDistributor[T]) distributeEvent(event T) {
+	a.consumerLock.RLock()
+	defer a.consumerLock.RUnlock()
+	for _, consumer := range a.consumers {
+		consumer.ConsumeEvent(event)
+	}
+}
+
+// Submit is the main entry point for the AsyncEventDistributor.
+// It receives an event, and asynchronously queues it for distribution.
+// The event submission is blocking, but the actual distribution is asynchronous and non-blocking. The event is queued up
+// in the message store, and the distributor worker picks it up from there and distributes it to all the consumers.
+// On a happy path Submit returns nil. It returns ErrSubmissionFailed if the event could not be queued up for processing, e.g.,
+// due to the message store being full.
+// It is safe to call Submit concurrently.
+// It is safe to call Submit after Shutdown.
+func (a *AsyncEventDistributor[T]) Submit(event T) error {
```

Comment:
I think the error return could be replaced by a boolean.

Reply:
In the worker-pool version `Submit` returns boolean as you suggested: 

https://github.com/onflow/flow-go/blob/e6ad0177edb567886831bfa34271a7295660fa4e/engine/common/worker/worker.go#L45-L48